app/services/base_service.rb

```diff
@@ -0,0 +1,90 @@
+module BaseService
+  CROP = 'Cannabis'.freeze
+
+  SEEDING_UNIT_MAP = {
+    'testing_package' => 'package',
+    'plant_barcoded' => 'plant',
+    'plants_barcoded' => 'plant',
+    'plant_clone' => 'plant',
+    'plants_clone' => 'plant',
+    'clones' => 'plant',
+    'clone' => 'plant',
+    'plants' => 'plant',
+    'plant' => 'plant'
+  }.freeze
+
+  WEIGHT_UNIT_MAP = {
+    'mg' => 'Milligrams',
+    'Milligram' => 'Milligrams',
+    'g' => 'Grams',
+    'Gram' => 'Grams',
+    'kg' => 'Kilograms',
+    'Kilogram' => 'Kilograms',
+    'oz' => 'Ounces',
+    'Ounce' => 'Ounces',
+    'lb' => 'Pounds',
+    'Pound' => 'Pounds'
+  }.freeze
+
+  def perform_action(ctx, integration, task = nil)
+    Lookup.new(ctx, integration, task).perform_action
+  end
+
+  def run_now?(ctx, integration)
+    Lookup.new(ctx, integration).run_mode == :now
+  end
+
+  def self.included(base)
+    base.send(:module_function, :perform_action, :run_now?)
+  end
+
+  class Lookup
+    def initialize(ctx, integration, task = nil)
+      @ctx = ctx
+      @integration = integration
+      @task = task
+    end
+
+    delegate :seeding_unit, to: :batch
+    delegate :run_mode, to: :module_for_completion
+
+    def perform_action
+      handler = module_for_completion
+
+      @task&.current_action = handler.name.underscore
+      handler.call(@ctx, @integration, batch)
+    end
+
+    def module_for_completion
+      action_type = completion.action_type.camelize
+      seeding_unit_name = module_name_for_seeding_unit.camelize
+
+      @integration.vendor_module.const_get("#{seeding_unit_name}::#{action_type}")
+    rescue NameError
+      raise InvalidOperation, "Processing not supported for #{seeding_unit.name} #{action_type} completions"
+    end
+
+    private
+
+    def module_name_for_seeding_unit
+      name = seeding_unit.name.parameterize(separator: '_')
+      SEEDING_UNIT_MAP.fetch(name, name)
+    end
+
+    def batch
```

Comment:
It would be awesome if we could query unrelated batches to the completion we received. It would be useful for those consume completions.

Reply:
I'll put something in my work getting NCS inline with these changes. Tracking this in #80